File: scripts/import-data.py
#!/usr/bin/env python3
"""
Script d'import automatique des données CSV vers DuckDB
Usage: python3 scripts/import-data.py
"""

import duckdb
import os
from pathlib import Path
import sys
from datetime import datetime

# Chemins
SCRIPT_DIR = Path(__file__).parent
DATA_DIR = Path("/Users/marceau/Desktop/test data/data/nouveaux")
DB_FILE = SCRIPT_DIR.parent / "public" / "duckdb.db"

FILES = {
    "clients": DATA_DIR / "client.csv",
    "produits": DATA_DIR / "Produits.csv",
    "magasins": DATA_DIR / "Points de vente.csv",
    "transactions": DATA_DIR / "détail transactions.csv"
}

def print_progress(msg):
    """Affiche un message avec timestamp"""
    timestamp = datetime.now().strftime("%H:%M:%S")
    print(f"[{timestamp}] {msg}")

def verify_files():
    """Vérifie que tous les fichiers existent"""
    print_progress("📁 Vérification des fichiers...")
    for name, path in FILES.items():
        if not path.exists():
            print(f"❌ Fichier manquant: {path}")
            sys.exit(1)
        size = path.stat().st_size / 1024 / 1024
        print(f"   ✓ {path.name} ({size:.2f} MB)")
    print()

def create_database():
    """Crée la base de données et se connecte"""
    print_progress("🗄️  Création de la base de données...")
    
    # Supprimer l'ancienne base si elle existe
    if DB_FILE.exists():
        DB_FILE.unlink()
        print_progress("   Base existante supprimée")
    
    conn = duckdb.connect(str(DB_FILE))
    print_progress("   ✅ Base créée")
    return conn

def import_clients(conn):
    """Importe les clients"""
    print_progress("👥 Import des clients...")
    
    try:
        # Créer la table en normalisant les noms de colonnes
        conn.execute(f"""
            CREATE TABLE clients AS 
            SELECT 
                "N° Carte fidélité" as carte,
                "Nom adresse" as nom,
                "Nom adresse" as prenom,
                "Sexe" as sexe,
                "Date de naissance" as date_naissance,
                "C.P" as cp,
                "Ville" as ville
            FROM read_csv_auto('{FILES["clients"]}', 
                delim=';', 
                header=true,
                ignore_errors=true,
                sample_size=-1
            )
        """)
        
        count = conn.execute("SELECT COUNT(*) FROM clients").fetchone()[0]
        print_progress(f"   ✅ {count:,} clients importés")
        return count
        
    except Exception as e:
        print_progress(f"   ❌ Erreur: {e}")
        raise

def import_produits(conn):
    """Importe les produits"""
    print_progress("📦 Import des produits...")
    
    try:
        conn.execute(f"""
            CREATE TABLE produits AS 
            SELECT 
                "N° Produit" as id,
                "Famille" as label,
                "Famille" as famille,
                "Sous famille" as sous_famille,
                "Sous sous famille" as sous_sous_famille,
                "Sous sous sous famille" as sous_sous_sous_famille
            FROM read_csv_auto('{FILES["produits"]}', 
                delim=';', 
                header=true,
                ignore_errors=true
            )
        """)
        
        count = conn.execute("SELECT COUNT(*) FROM produits").fetchone()[0]
        print_progress(f"   ✅ {count:,} produits importés")
        return count
        
    except Exception as e:
        print_progress(f"   ❌ Erreur: {e}")
        raise

def import_magasins(conn):
    """Importe les magasins"""
    print_progress("🏪 Import des magasins...")
    
    try:
        conn.execute(f"""
            CREATE TABLE magasins AS 
            SELECT 
                "N° Dépôt" as code,
                "Intitulé dépôt" as nom,
                "Zones magasin" as zone,
                "Adresse 1" as ville
            FROM read_csv_auto('{FILES["magasins"]}', 
                delim=';', 
                header=true,
                ignore_errors=true
            )
        """)
        
        count = conn.execute("SELECT COUNT(*) FROM magasins").fetchone()[0]
        print_progress(f"   ✅ {count:,} magasins importés")
        return count
        
    except Exception as e:
        print_progress(f"   ❌ Erreur: {e}")
        raise

def import_transactions(conn):
    """Importe les transactions avec progression"""
    print_progress("🛒 Import des transactions (ceci peut prendre plusieurs minutes)...")
    
    try:
        start = datetime.now()
        
        conn.execute(f"""
            CREATE TABLE transactions AS 
            SELECT 
                "N° Facture client" as facture,
                "Date facture" as date,
                "N° Carte fidélité" as carte,
                "N° Produit" as produit,
                CAST(REPLACE("Quantité unitaire", ',', '.') as DOUBLE) as quantite,
                CAST(REPLACE("Quantité unitaire", ',', '.') as DOUBLE) * 
                CAST(REPLACE("Prix vente net en devise société", ',', '.') as DOUBLE) as ca,
                "Dépôt" as depot,
                false as is_web
            FROM read_csv_auto('{FILES["transactions"]}', 
                delim=';', 
                header=true,
                ignore_errors=true,
                sample_size=-1
            )
        """)
        
        count = conn.execute("SELECT COUNT(*) FROM transactions").fetchone()[0]
        duration = (datetime.now() - start).total_seconds()
        print_progress(f"   ✅ {count:,} transactions importées en {duration:.0f}s")
        return count
        
    except Exception as e:
        print_progress(f"   ❌ Erreur: {e}")
        raise

def create_indexes(conn):
    """Crée les indexes pour améliorer les performances"""
    print_progress("📊 Création des indexes...")
    
    try:
        # Déterminer les noms de colonnes réels
        cols = [row[0] for row in conn.execute("PRAGMA table_info(transactions)").fetchall()]
        
        # Créer les indexes sur les colonnes qui existent
        if any('date' in c.lower() for c in cols):
            date_col = next(c for c in cols if 'date' in c.lower())
            conn.execute(f'CREATE INDEX idx_transactions_date ON transactions("{date_col}")')
        
        print_progress("   ✅ Indexes créés")
        
    except Exception as e:
        print_progress(f"   ⚠️  Erreur indexes (non bloquant): {e}")

def show_summary(conn, stats):
    """Affiche un résumé des données importées"""
    print_progress("\n" + "="*60)
    print_progress("🎉 IMPORT TERMINÉ AVEC SUCCÈS!")
    print_progress("="*60)
    print(f"\n📊 Résumé:")
    print(f"   • Clients:      {stats['clients']:>10,}")
    print(f"   • Produits:     {stats['produits']:>10,}")
    print(f"   • Magasins:     {stats['magasins']:>10,}")
    print(f"   • Transactions: {stats['transactions']:>10,}")
    print(f"\n💾 Base de données: {DB_FILE}")
    print(f"   Taille: {DB_FILE.stat().st_size / 1024 / 1024:.2f} MB")
    
    # Vérifier le CA total
    ca_result = conn.execute("SELECT SUM(ca) as ca_total FROM transactions").fetchone()
    ca_total = ca_result[0]
    print(f"\n💰 CA Total: {ca_total:,.2f} €")
    
    # Afficher les colonnes des tables
    print("\n📋 Structure des tables:")
    for table in ['clients', 'produits', 'magasins', 'transactions']:
        cols = conn.execute(f"PRAGMA table_info({table})").fetchall()
        print(f"\n   {table.upper()}:")
        for col in cols[:5]:  # Premières 5 colonnes
            print(f"      - {col[1]} ({col[2]})")
        if len(cols) > 5:
            print(f"      ... et {len(cols)-5} autres colonnes")
    
    print("\n✅ Vous pouvez maintenant rafraîchir le navigateur!")
    print()

def main():
    """Fonction principale"""
    print()
    print("="*60)
    print("🚀 IMPORT AUTOMATIQUE DES DONNÉES CSV VERS DUCKDB")
    print("="*60)
    print()
    
    start_time = datetime.now()
    stats = {}
    
    try:
        # 1. Vérifier les fichiers
        verify_files()
        
        # 2. Créer la base
        conn = create_database()
        
        # 3. Importer les données
        stats['clients'] = import_clients(conn)
        stats['produits'] = import_produits(conn)
        stats['magasins'] = import_magasins(conn)
        stats['transactions'] = import_transactions(conn)
        
        # 4. Créer les indexes
        create_indexes(conn)
        
        # 5. Afficher le résumé
        duration = (datetime.now() - start_time).total_seconds()
        print_progress(f"\n⏱️  Durée totale: {duration:.0f}s ({duration/60:.1f}min)")
        show_summary(conn, stats)
        
        conn.close()
        sys.exit(0)
        
    except Exception as e:
        print()
        print_progress(f"❌ ERREUR FATALE: {e}")
        print()
        import traceback
        traceback.print_exc()
        sys.exit(1)

if __name__ == "__main__":
    main()

File: dev/scripts/import-data.js
#!/usr/bin/env node

/**
 * Script d'import automatique des données CSV vers DuckDB
 * Usage: node scripts/import-data.js
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { execSync } from 'child_process';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Chemins des fichiers - tous dans "data/nouveaux/"
const DATA_DIR = path.resolve(__dirname, '../../data/nouveaux');

const FILES = {
  clients: path.join(DATA_DIR, 'client.csv'),
  produits: path.join(DATA_DIR, 'Produits.csv'),
  magasins: path.join(DATA_DIR, 'Points de vente.csv'),
  tickets: path.join(DATA_DIR, 'détail transactions.csv')
};

const DB_FILE = path.join(__dirname, '../public/duckdb.db');

// Vérifier que DuckDB Node est installé
let duckdb;
try {
  duckdb = (await import('duckdb')).default;
  console.log('✅ DuckDB Node détecté');
} catch (e) {
  console.error('❌ DuckDB Node non installé. Installation...');
  execSync('npm install duckdb', { stdio: 'inherit' });
  duckdb = (await import('duckdb')).default;
}

// Statistiques
const stats = {
  startTime: Date.now(),
  clients: 0,
  produits: 0,
  magasins: 0,
  transactions: 0,
  errors: []
};

async function main() {
  console.log('🚀 Démarrage import automatique...\n');

  // Vérifier les fichiers
  console.log('📁 Vérification des fichiers...');
  for (const [name, filePath] of Object.entries(FILES)) {
    if (!fs.existsSync(filePath)) {
      console.error(`❌ Fichier manquant: ${filePath}`);
      process.exit(1);
    }
    const size = (fs.statSync(filePath).size / 1024 / 1024).toFixed(2);
    console.log(`   ✓ ${path.basename(filePath)} (${size} MB)`);
  }

  // Créer la base de données
  console.log('\n🗄️  Création base de données...');
  const db = new duckdb.Database(DB_FILE);
  const conn = db.connect();

  try {
    // Créer le schéma
    await createSchema(conn);
    
    // Importer les données
    await importClients(conn);
    await importProduits(conn);
    await importMagasins(conn);
    await importTransactions(conn);
    
    // Générer le snapshot
    await generateSnapshot(conn);
    
    // Afficher les résultats
    printResults();
    
  } catch (error) {
    console.error('\n❌ Erreur fatale:', error.message);
    stats.errors.push(error.message);
    process.exit(1);
  } finally {
    conn.close();
    db.close();
  }
}

function createSchema(conn) {
  return new Promise((resolve, reject) => {
    console.log('   Création tables...');
    
    const sql = `
      CREATE TABLE IF NOT EXISTS clients_raw (
        carte VARCHAR,
        date_creation VARCHAR,
        statut VARCHAR,
        date_validite VARCHAR,
        civilite VARCHAR,
        date_naissance VARCHAR,
        sexe VARCHAR,
        nom_adresse VARCHAR,
        adresse1 VARCHAR,
        adresse2 VARCHAR,
        adresse3 VARCHAR,
        cp VARCHAR,
        ville VARCHAR
      );
      
      CREATE TABLE IF NOT EXISTS produits (
        id VARCHAR PRIMARY KEY,
        famille VARCHAR,
        sous_famille VARCHAR,
        sous_sous_famille VARCHAR,
        sous_sous_sous_famille VARCHAR
      );
      
      CREATE TABLE IF NOT EXISTS magasins_raw (
        zone VARCHAR,
        code VARCHAR,
        nom VARCHAR,
        adresse1 VARCHAR,
        adresse2 VARCHAR,
        adresse3 VARCHAR,
        cp VARCHAR,
        ville VARCHAR
      );
      
      CREATE TABLE IF NOT EXISTS transactions_raw (
        carte VARCHAR,
        facture VARCHAR,
        depot VARCHAR,
        date VARCHAR,
        produit VARCHAR,
        quantite VARCHAR,
        ca VARCHAR
      );
      
      CREATE TABLE IF NOT EXISTS dashboard_snapshot (
        year INTEGER PRIMARY KEY,
        payload TEXT
      );
      
      CREATE TABLE IF NOT EXISTS metadata (
        key VARCHAR PRIMARY KEY,
        value TEXT,
        updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
      );
    `;
    
    conn.exec(sql, (err) => {
      if (err) reject(err);
      else {
        console.log('   ✅ Tables créées');
        resolve();
      }
    });
  });
}

async function importClients(conn) {
  return new Promise((resolve, reject) => {
    const file = FILES.clients;
    console.log('\n👥 Import clients...');
    
    conn.exec(`
      COPY clients_raw FROM '${file}' (HEADER TRUE, DELIMITER ';', IGNORE_ERRORS TRUE);
      SELECT COUNT(*) as count FROM clients_raw;
    `, (err, result) => {
      if (err) {
        stats.errors.push(`Clients: ${err.message}`);
        reject(err);
      } else {
        // exec retourne un tableau de résultats, prendre le dernier (SELECT COUNT)
        const countResult = result[result.length - 1];
        stats.clients = countResult.count;
        console.log(`   ✅ ${stats.clients.toLocaleString()} clients importés`);
        resolve();
      }
    });
  });
}

async function importProduits(conn) {
  return new Promise((resolve, reject) => {
    const file = FILES.produits;
    console.log('\n📦 Import produits...');
    
    conn.exec(`
      COPY produits FROM '${file}' (HEADER TRUE, DELIMITER ';', IGNORE_ERRORS TRUE);
      SELECT COUNT(*) as count FROM produits;
    `, (err, result) => {
      if (err) {
        stats.errors.push(`Produits: ${err.message}`);
        reject(err);
      } else {
        const countResult = result[result.length - 1];
        stats.produits = countResult.count;
        console.log(`   ✅ ${stats.produits.toLocaleString()} produits importés`);
        resolve();
      }
    });
  });
}

async function importMagasins(conn) {
  return new Promise((resolve, reject) => {
    const file = FILES.magasins;
    console.log('\n🏪 Import magasins...');
    
    conn.exec(`
      COPY magasins_raw FROM '${file}' (HEADER TRUE, DELIMITER ';', IGNORE_ERRORS TRUE);
      SELECT COUNT(*) as count FROM magasins_raw;
    `, (err, result) => {
      if (err) {
        stats.errors.push(`Magasins: ${err.message}`);
        reject(err);
      } else {
        const countResult = result[result.length - 1];
        stats.magasins = countResult.count;
        console.log(`   ✅ ${stats.magasins.toLocaleString()} magasins importés`);
        resolve();
      }
    });
  });
}

async function importTransactions(conn) {
  return new Promise((resolve, reject) => {
    const file = FILES.tickets;
    console.log('\n🧾 Import transactions...');
    console.log('   (Cela peut prendre plusieurs minutes pour 6M lignes...)');
    
    const startTime = Date.now();
    
    conn.exec(`
      COPY transactions_raw FROM '${file}' (HEADER TRUE, DELIMITER ';', IGNORE_ERRORS TRUE);
      SELECT COUNT(*) as count FROM transactions_raw;
    `, (err, result) => {
      if (err) {
        stats.errors.push(`Transactions: ${err.message}`);
        reject(err);
      } else {
        const duration = ((Date.now() - startTime) / 1000).toFixed(1);
        const countResult = result[result.length - 1];
        stats.transactions = countResult.count;
        console.log(`   ✅ ${stats.transactions.toLocaleString()} transactions importées (${duration}s)`);
        resolve();
      }
    });
  });
}

async function generateSnapshot(conn) {
  return new Promise((resolve, reject) => {
    console.log('\n📸 Génération snapshot 2025...');
    
    conn.all(`
      WITH year_data AS (
        SELECT 
          t.*,
          p.famille,
          p.sous_famille,
          m.nom as magasin_nom,
          m.zone as magasin_zone,
          c.sexe,
          c.cp as client_cp,
          c.ville as client_ville
        FROM transactions t
        LEFT JOIN produits p ON t.produit = p.id
        LEFT JOIN magasins m ON t.depot = m.code
        LEFT JOIN clients c ON t.carte = c.carte
        WHERE t.date >= '2025-01-01' AND t.date <= '2025-12-31'
      )
      SELECT
        COUNT(*) as total_transactions,
        COUNT(DISTINCT carte) as total_clients,
        COUNT(DISTINCT facture) as total_tickets,
        SUM(ca) as ca_total,
        SUM(CASE WHEN is_web = false THEN ca ELSE 0 END) as ca_mag,
        SUM(CASE WHEN is_web = true THEN ca ELSE 0 END) as ca_web,
        COUNT(CASE WHEN is_web = false THEN 1 END) as trans_mag,
        COUNT(CASE WHEN is_web = true THEN 1 END) as trans_web
      FROM year_data
    `, (err, result) => {
      if (err) {
        stats.errors.push(`Snapshot: ${err.message}`);
        reject(err);
      } else {
        const row = result[0];
        const payload = JSON.stringify({
          year: 2025,
          stats: {
            total_transactions: Number(row.total_transactions),
            total_clients: Number(row.total_clients),
            total_tickets: Number(row.total_tickets),
            ca_total: Number(row.ca_total),
            ca_mag: Number(row.ca_mag),
            ca_web: Number(row.ca_web),
            trans_mag: Number(row.trans_mag),
            trans_web: Number(row.trans_web)
          },
          generated_at: new Date().toISOString()
        });
        
        const escapedPayload = payload.replace(/'/g, "''");
        
        conn.exec(`
          INSERT OR REPLACE INTO dashboard_snapshot (year, payload)
          VALUES (2025, '${escapedPayload}')
        `, (err) => {
          if (err) {
            stats.errors.push(`Save snapshot: ${err.message}`);
            reject(err);
          } else {
            console.log('   ✅ Snapshot généré');
            resolve();
          }
        });
      }
    });
  });
}

function printResults() {
  const duration = ((Date.now() - stats.startTime) / 1000).toFixed(1);
  
  console.log('\n' + '='.repeat(60));
  console.log('🎉 IMPORT TERMINÉ AVEC SUCCÈS');
  console.log('='.repeat(60));
  console.log(`\n📊 Statistiques:`);
  console.log(`   • Clients:       ${stats.clients.toLocaleString()}`);
  console.log(`   • Produits:      ${stats.produits.toLocaleString()}`);
  console.log(`   • Magasins:      ${stats.magasins.toLocaleString()}`);
  console.log(`   • Transactions:  ${stats.transactions.toLocaleString()}`);
  console.log(`\n⏱️  Temps total: ${duration}s`);
  console.log(`\n💾 Base de données: ${DB_FILE}`);
  console.log(`\n✨ Prêt à utiliser! Rafraîchis ton navigateur.\n`);
}

// Lancer le script
main().catch(err => {
  console.error('\n💥 Erreur:', err);
  process.exit(1);
});
